docs/hugo/content/design/ADR-2022-09-Reading-Status-Properties-Of-Resources.md

```diff
@@ -0,0 +1,261 @@
+```
+
+or (for the more generic proposals)
+```yaml
+RoleAssignmentProperties:
+  PrincipalId:
+    $arbitraryReference: true
+```
+
+### Option 1: Support a new type similar to ResourceReference, but specifically for AAD references
+
+That would look something like:
+```go
+// +kubebuilder:object:generate=true
+type IdentityReference struct {
+
+    // Name of the identity in Kuberentes
+    Name string `json:"name,omitempty"`
+
+    // PrincipalId of the identity in AAD. This is mutually exclusive with Name.
+    PrincipalId string `json:"principalId,omitempty"`
+}
+```
+
+Usage would be something like:
+```yaml
+apiVersion: managedidentity.azure.com/v1beta20181130
+kind: UserAssignedIdentity
+metadata:
+  name: sample-uai
+spec:
+  location: Germany West Central
+  owner:
+    name: dev-sample-rg
+---
+apiVersion: authorization.azure.com/v1beta20200801preview
+kind: RoleAssignment
+metadata:
+  name: 6a2d44f5-57d8-4916-9f46-ff7c9c1b338f
+spec:
+  location: Germany West Central
+  owner:
+    name: samplevnet
+    group: network.azure.com
+    kind: VirtualNetwork
+  principalId:
+    name: sample-uai
+  roleDefinitionReference:
+    armId: /subscriptions/00000000-0000-0000-0000-000000000000/providers/Microsoft.Authorization/roleDefinitions/b24988ac-6180-42a0-ab88-20f7382dd24c
+```
+
+
+Pros:
+
+1. Relatively easy to implement
+2. Simple to use
+3. Can give a good error message if the identity in question doesn't exist
+
+Cons:
+
+1. Not generic. What if some resources require `ClientId` instead of `PrincipalId`? We would need a new type for that.
+
+
+### Option 2: Support a generic resource for reading from other ASO resources Status
+
+```go
+// +kubebuilder:object:generate=true
+type StringReference struct {
+
+    Value string `json:"value,omitempty"`
+
+    Ref *StringReferenceRef `json:"ref,omitempty"`
+}
+
+// +kubebuilder:object:generate=true
+type StringReferenceRef struct {
+    KubernetesResourceReference
+
+    Path string `json:"path,omitempty"`
+}
+
+type KubernetesResourceReference struct {
+    // Group is the Kubernetes group of the resource.
+    Group string `json:"group,omitempty"`
+    // Kind is the Kubernetes kind of the resource.
+    Kind string `json:"kind,omitempty"`
+    // Name is the Kubernetes name of the resource.
+    Name string `json:"name,omitempty"`
+}
+```
+
+Usage would be something like:
+```yaml
+apiVersion: managedidentity.azure.com/v1beta20181130
+kind: UserAssignedIdentity
+metadata:
+  name: sample-uai
+spec:
+  location: Germany West Central
+  owner:
+    name: dev-sample-rg
+---
+apiVersion: authorization.azure.com/v1beta20200801preview
+kind: RoleAssignment
+metadata:
+  name: 6a2d44f5-57d8-4916-9f46-ff7c9c1b338f
+spec:
+  location: Germany West Central
+  owner:
+    name: samplevnet
+    group: network.azure.com
+    kind: VirtualNetwork
+  principalId:
+    ref:
+      name: sample-uai
+      group: manangedidentity.azure.com
+      kind: UserAssignedIdentity
+      path: status.principalId
```

Comment:
Worth observing (perhaps in a separate `Implementation` section below `Pros` & `Cons`) that evaluation of an arbitrary path has some challenges - we could code it ourselves, and handle all the edge cases, or take a shortcut 
(perhaps by serialization to JSON and using a query library) at the risk of not having control over some cases.

Reply:
Also make a note that every ASO resource would have to register for events related to every other ASO resource, resulting in a significant load on the Kubernetes Event system, even for workloads that don't use this feature.

A good principle to reference here is "Pay for Play" - if someone isn't using a particular ASO feature, they shouldn't have to pay very much (if anything) for it's unused presence.

Reply:
Because there would be literally no restriction on which property a user might reference from another resource's status, it would be possible for one resource to consume a volatile property from another (say, CPU load from a VM).

If a user did this, every reconcile would result in an update to the resource, which might trigger a busy wait loop and a nasty failure mode.

Reply:
> Worth observing (perhaps in a separate Implementation section below Pros & Cons) that evaluation of an arbitrary path has some challenges - we could code it ourselves, and handle all the edge cases, or take a shortcut
(perhaps by serialization to JSON and using a query library) at the risk of not having control over some cases.

I've just put this into the existing con, as I think it's a con and makes more sense there. I've added `Implementation` as a section in a few other places where there are "notes" on implementation that aren't strictly a pro/con.